src/game/AI/PlayerAI/PlayerAI.cpp

```diff
@@ -102,15 +102,27 @@ void PlayerAI::ExecuteSpells()
 
 void PlayerAI::JustGotCharmed(Unit* charmer)
 {
-    m_player->GetMotionMaster()->MoveFollow(charmer, PET_FOLLOW_DIST, PET_FOLLOW_ANGLE, true);
+    if (charmer->GetFormationSlot())
+    {
+        charmer->GetFormationSlot()->GetFormationData()->Add(m_player);
+    }
+    else
+        m_player->GetMotionMaster()->MoveFollow(charmer, PET_FOLLOW_DIST, PET_FOLLOW_ANGLE, true);
     AttackClosestEnemy();
 }
 
 void PlayerAI::EnterEvadeMode()
 {
     m_player->CombatStopWithPets(true);
     if (Unit* charmer = m_player->GetCharmer())
-        m_player->GetMotionMaster()->MoveFollow(charmer, PET_FOLLOW_DIST, PET_FOLLOW_ANGLE, true);
+    {
+        if (charmer->GetFormationSlot())
+        {
+            charmer->GetFormationSlot()->GetFormationData()->Add(m_player);
+        }
+        else
+            m_player->GetMotionMaster()->MoveFollow(charmer, PET_FOLLOW_DIST, PET_FOLLOW_ANGLE, true);
```

Comment:
Since there is multiple of these, maybe handle it in MoveFollow?

Reply:
Well first at this stage formation is a sub option for creaturegroup meaning that formation need a creature group absolutly and its designed only for creature as leader.
Thus some big change have to be done to be able to make a player as a master and more to have some other kind of group that will be attached to player.

Secondly you are asking for more than just adding it as exception of the movefollow we had a discussion about it with tobi and it seem that movefollow should even been replaced by some 'move in formation' with position computed depending of the player's followers.